File: Homework_1/matrix.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

// Multiply two matrix
void mul_matrix(int n, long long matrix1[][n], long long matrix2[][n], long long matrix[][n]) {
    for(int i = 0; i < n; ++i) {
        for(int j = 0; j < n; ++j) {
            matrix[i][j] = 0;
        }
    }

    int i = 0, h = 0;

    while(i != n) {
        int temp = 0;
        for(int j = 0; j < n; ++j) {
            temp += matrix1[i][j] * matrix2[j][h];
        }
        matrix[i][h] = temp;
        h++;
        if(h == n) {
            h = 0;
            i++;
        }
    }
}

double solve() {
    FILE *fp = NULL;

    // Open file 
    fp = fopen("generate_matrix.txt", "r");

    // Read size of matrix
    int n;
    fscanf(fp, "%d\n", &n);

    long long matrix1[n][n], matrix2[n][n], matrix[n][n];

    // Read matrix A
    for(int i = 0; i < n; ++i) {
        for(int j = 0; j < n; ++j) {
            fscanf(fp, "%d ", &matrix1[i][j]);
        }
    }
    
    // Read matrix B
    for(int i = 0; i < n; ++i) {
        for(int j = 0; j < n; ++j) {
            fscanf(fp, "%d ", &matrix2[i][j]);
        }
    }

    clock_t start, end;
    // Record start time
    start = clock();

    mul_matrix(n, matrix1, matrix2, matrix);

    // Record end time
    end = clock();

    // Calculate runtime of function multiply two matrix
    double duration = ((double)end - start) / ((double)CLOCKS_PER_SEC);

    fclose(fp);
    
    return duration;
}

int main() {

    double duration = solve();
    
    printf("%0.10f", duration);

    return 0;
}

File: Homework_1/matrix.py
import random as rd
import time as t

def generate_matrix(n):
    matrix = []
    
    for i in range(n):
        matrix.append([])
        for j in range(n):
            matrix[i].append(rd.randint(-10, 10))
    
    return matrix

def mul_matrix(n, m1, m2):
    m = []
    
    for i in range(n):
        m.append([])
        for j in range(n):
            m[i].append(0)
            
    i, h = 0, 0
    
    while i != n:
        temp = 0
        for j in range(n):
            temp += m1[i][j] * m2[j][h]        
        m[i][h] = temp
        h += 1
        if h == n:
            i += 1
            h = 0
    
    return m

# Random size of square matrix
n = rd.randint(100, 500)

# Generate matrix A and B with value in range -10 -> 10
matrix1 = generate_matrix(n)
matrix2 = generate_matrix(n)

# Record start time
start = t.time()

# Multiplication matrix
matrix = mul_matrix(n, matrix1, matrix2)

# Record end time
end = t.time()

# Calculate runtime of function multiply two matrix
print("Runtime : ", end = "")
print("{0:.10f}".format((end - start) * 10**3))